typescript/search-attributes.md

```diff
@@ -0,0 +1,99 @@
+## Search attributes V2
+
+### Why do we need this?
+
+In the current version of the SDK, search attributes are always arrays. In order for the server to support visibility
+on top of standard persistence, we need to redo the search attributes API to enforce single values for all search
+attributes except for keywords. The server will eventually drop support for multiple search attribute values for
+non-keywords.
+
+### Minimal proposal
+
+Make the E2E experience consistent, including query construction for listing workflows and search attribute
+registration.
+
+`workflows.ts`
+
+```ts
+import { defineAttribute, Keyword, Text } from '@temporalio/common/search-attributes';
+
+export const tagsAttribute = defineAttribute(Keyword, 'tags');
+export const customerAttribute = defineAttribute(Text, 'customer');
+export const dateOfPurchaseAttribute = defineAttribute(Date, 'dateOfPurchase');
+
+export async function myWorkflow() {
+  // Get the value of a search attribute - returns string | undefined here
+  const customer = customerAttribute.value();
+  // Get the value of a search attribute - returns string[] (which will be empty if unset)
+  const tags = tagsAttribute.value();
+  // ... random workflow logic ...
+
+  // Set the value of a search attribute, all set and unset calls are buffered and flushed out as a single command at
+  // the end of an activation
+  dateOfPurchaseAttribute.set(new Date());
+  // Explicit unset API
+  dateOfPurchaseAttribute.unset();
+  // ...
+}
+```
+
+`client.ts`
+
+```ts
+import { WorkflowType } from '@temporalio/common/search-attributes';
+import { Client, Q } from '@temporalio/client';
+import { customerAttribute, dateOfPurchaseAttribute, myWorkflow, tagsAttribute } from './workflows';
+
+const client = new Client(/* options */);
+
+// Only available in OSS for now
+await client.operator.registerSearchAttributes(customerAttribute, dateOfPurchaseAttribute);
+// Other methods will be exposed too
+
+const iterator = client.workflow.list({
```

Comment:
Quick question as I start to work on this in Python. So this solves https://github.com/temporalio/sdk-features/issues/75 too? I assume this is interceptable? Also, can you show a quick example of status filtering? Is `ExecutionStatus` a search attribute?

Reply:
This will solve the high level SDK list API, assuming you want to add the query builder (you should IMHO).
I don't see much value in making it intercept-able if there are no headers to pass to the server.
`ExecutionStatus` along with the rest of the [default search attributes](https://docs.temporal.io/visibility#default-search-attributes) will be defined in the SDK.

For search attributes that we know are enums I'd definitely restrict the type (I think it's only ExecutionStatus).

Something like:
```ts
Q.filter(builtinSearchAttributes.ExecutionStatus.eq('RUNNING'))
```

Reply:
>  I don't see much value in making it intercept-able if there are no headers to pass to the server.

While I thought similar, I was overridden on workflow handle "describe". I think we are now forced to be consistent here.